language/internal/cc/parser/parser.go

```diff
@@ -267,145 +257,153 @@ func parseDefinedExpr(p *parser, op string) (Expr, error) {
-					}
-					break
-				}
-				// parseDirective assumes full directive name including '#' prefix
-				token = "#" + directiveKind
-			}
-			directive, err := p.parseDirective(token)
-			if err != nil {
-				skipped, _ := p.readUntilEOL() // skip remaining part of directive
-				if debug {
-					log.Printf("Failed to parse %v directive: %v, skipping tokens until end of line: %v", token, err, skipped)
-				}
-				break
-			}
-			directives = append(directives, directive)
+// Return the next token and consume it, or TokenEmpty if no tokens are left.
+func (p *parser) nextToken() lexer.Token {
+	if len(p.tokensLeft) == 0 {
+		return lexer.TokenEOF
+	}
 
-		case token == "main":
-			if next, exists := p.tr.next(); exists && next == "(" {
-				if prev == "int" {
-					p.sourceInfo.HasMain = true
+	token := p.tokensLeft[0]
+	p.dropTokens(1)
+	return token
+}
+
+// Check if the next token matches the expected content, returning error
+// otherwise.
+func (p *parser) expectNextToken(expected lexer.TokenType) error {
+	tokenType := p.nextToken().Type
+	switch tokenType {
+	case lexer.TokenType_EOF:
+		return fmt.Errorf("expected %q but reached end of input", expected)
+	case expected:
+		return nil
+	default:
+		return fmt.Errorf("expected %q but found %q", expected, tokenType)
+	}
+}
+
+// parseDirectivesUntil reads tokens and parses directives until shouldStop
+// returns true. It handles main(), #include, and preprocessor blocks, and
+// builds the nested directive structure.
+func (p *parser) parseDirectivesUntil(shouldStop func(token lexer.TokenType) bool) ([]Directive, error) {
+	directives := []Directive{}
+	for !shouldStop(p.peekToken()) {
+		p.parseMainFunction()
+
+		if tokenType := p.nextToken().Type; tokenType.IsPreprocessorDirective() {
+			directive, err := p.parseDirective(tokenType)
+			if err == nil {
+				directives = append(directives, directive)
+			} else {
+				skipped := p.readUntilEOL() // skip remaining part of directive
+				if debug {
+					log.Printf("Failed to parse %v directive: %v, skipping tokens until end of line: %v", tokenType, err, skipped)
 				}
 			}
 		}
 	}
+
+	return directives, nil
 }
 
-// parseExpr parses a preprocessor expression (#if/#elif condition) as an Expr AST.
+// parseExpr parses a preprocessor expression (#if/#elif condition) as an Expr
+// AST.
 func (p *parser) parseExpr() (Expr, error) {
 	return p.parseExprPrecedence(precedenceLowest)
 }
 
-// nextToken returns the next token or an error if EOF is reached.
-func (p *parser) nextToken() (string, error) {
-	token, ok := p.tr.next()
-	if !ok {
-		return "", errors.New("expected identifier, found EOF")
-	}
-	if token == EOL {
-		return "", errors.New("expected token, found EOL")
+// Similar to nextToken(), but returns an error if no tokens are left or newline
+// is encountered. Newline means unexpected end of the directive in this
+// context.
+func (p *parser) nextDirectiveToken() (lexer.Token, error) {
```

Comment:
Rename to `expectDirectiveToken`. From the name alone, I'd expect `nextDirectiveToken` to skip tokens until it finds a directive.